File: 2015/Day8/src/day8.py
import sys

def count_in_memory(line: str):
    count = 0
    i = 1
    while(i < len(line) - 1): # ignore start " and end "
        count += 1
        if(line[i] == "\\" and i + 1 < len(line) - 1):
            if(line[i + 1] == "\""):
                i += 1
            elif(line[i + 1] == "\\"):
                i += 1
            elif(line[i + 1] == "x"):
                i += 3
        i += 1
    return count

if __name__ == "__main__":
    if(len(sys.argv) != 2):
        print("Usage: python3 day8.py input.txt")

    # part 1
    total_num_chars = 0
    in_memory_count = 0
    with open(sys.argv[1], "r") as fileobj:
        for line in fileobj:
            total_num_chars += len(line.rstrip())
            in_memory_count += count_in_memory(line.rstrip())
    print(f"Answer to part 1 is {total_num_chars - in_memory_count}")

    # part 2
    count = 0 
    with open(sys.argv[1], "r") as fileobj:
        for line in fileobj:
            line_count = 2 # starts with 2 chars, starting " and ending "
            i = 0
            while(i < len(line.rstrip())):
                line_count += 1
                if(line[i] == "\""):
                    line_count += 1
                if(line[i] == "\\"):
                    line_count += 1
                i += 1
            count += line_count
    print(f"Answer to part 2 is {count - total_num_chars}")

File: 2015/Day8/src/day8.cpp
#include <string>
#include <iostream>
#include <fstream>

#define DEBUG false

using namespace std;

int main(int argc, char * argv[])
{
    if(argc != 2)
    {
        cout << "Usage: ./day8 input.txt\n";
        exit(1);
    }
    ifstream file(argv[1]);
    string line;
    getline(file, line);
    int total_byte_count = 0;
    int in_memory_count = 0;
    int escaped_count = 0;
    while(line.size() > 0)
    {
        if(DEBUG)
        {
            cout << line << " " << line.size() << endl;
        }
        total_byte_count += line.size();
        int i = 1;
        //count in memory
        while(i < line.size() - 1)
        {
            in_memory_count += 1;
            if(line.at(i) == '\\' && i + 1 < line.size() - 1)
            {
                if(line.at(i + 1) == '\"')
                {
                    i += 1;
                }
                else if(line.at(i + 1) == 'x')
                {
                    i += 3;
                }
                else if(line.at(i + 1) == '\\')
                {
                    i += 1;
                }
            }
            i += 1;
        }
        i = 0;
        // count escaped
        escaped_count += 2; // extra two characters for the start and end "
        while(i < line.size())
        {
            escaped_count += 1;
            if(line.at(i) == '\\')
            {
                escaped_count += 1;
            }
            else if(line.at(i) == '\"')
            {
                escaped_count += 1;
            }
            i += 1;
        }
        if(DEBUG)
        {
            cout << in_memory_count << " " << escaped_count << endl;
        }
        getline(file, line);
    }
    file.close();

    cout << "The answer to part 1 is " << total_byte_count - in_memory_count << endl;

    cout << "The answer to part 2 is " << escaped_count - total_byte_count << endl;
}
